### PR TITLE
Add a ComputeReviewState function

### PR DESCRIPTION
**Pull request by mrnugget**
This adds a ComputeReviewState function that's to be used to compute the
review state.

It also brings back the ReviewState method on Changeset, except that
it's private now and only called for Bitbucket Server changesets.

### CODE REVIEW

File: internal/campaigns/types.go [hunk 1]
```diff
@@ -567,6 +567,37 @@ func (c *Changeset) Labels() []ChangesetLabel {
 	}
 }
 
+// ReviewState of a Changeset. GitHub doesn't keep the review state on a
+// changeset, so it a GitHub Changeset will always return
+// ChangesetReviewStatePending.
+// This method should not be called directly. Instead call
```
Comment (ryanslade) [inline internal/campaigns/types.go]: Missing comment

File: internal/campaigns/types.go [hunk 2]
```diff
@@ -567,6 +567,37 @@ func (c *Changeset) Labels() []ChangesetLabel {
 	}
 }
 
+// ReviewState of a Changeset. GitHub doesn't keep the review state on a
+// changeset, so it a GitHub Changeset will always return
```
Comment (ryanslade) [inline internal/campaigns/types.go]:
```suggestion
// changeset, so a GitHub Changeset will always return
```

File: enterprise/internal/campaigns/resolvers/resolver_test.go [hunk 3]
```diff
@@ -482,7 +482,7 @@ func TestCampaigns(t *testing.T) {
 					URL:         "https://bitbucket.sgdev.org/projects/SOUR/repos/vegeta/pull-requests/2",
 					ServiceType: "bitbucketServer",
 				},
-				ReviewState: "APPROVED",
+				ReviewState: "PENDING",
```
Comment (ryanslade) [inline enterprise/internal/campaigns/resolvers/resolver_test.go]: Does this mean our old code was incorrect?

Comment (mrnugget) [reply enterprise/internal/campaigns/resolvers/resolver_test.go]: Thanks for being so diligent! This was a nice catch. There _is_ a bug in the code. Currently investing.

Comment (mrnugget) [reply enterprise/internal/campaigns/resolvers/resolver_test.go]: Investi**gating** btw! I'm not investing in bugs.

**Comment by mrnugget** [comment]
@ryanslade Can you take another look? After you pointed it out, I dug into the tests and found out that we didn't handle Bitbucket's "Unapproved" event in the `(ChangesetEvents).ReviewState()` method. We _do_ handle it correct in the burndown chart. 
I added tests and fixed them by taking the event type into account.

I also added some mocks to the resolver tests so that they're less noisy.

File: internal/campaigns/types.go [hunk 4]
```diff
@@ -694,6 +726,28 @@ func ComputeChangesetState(c *Changeset, events ChangesetEvents) (ChangesetState
...
+
+	// GitHub only stores the ReviewState in events, we can't look at the
+	// Changeset.
+	if c.ExternalServiceType == github.ServiceType {
+		return events.reviewState()
+	}
+
+	// For other codehosts we check whether the Changeset is newer or the
+	// events and use the newest entity to get the reviewstate.
+	newestEvent := events[len(events)-1]
```
Comment (ryanslade) [inline internal/campaigns/types.go]: Maybe we should add a comment above the function specifying that events should be sorted

Comment (mrnugget) [reply internal/campaigns/types.go]: Like this?

https://github.com/sourcegraph/sourcegraph/pull/8978/files#diff-fa5e1bd8d3af52b49e4ee71fd894780eR730

Comment (ryanslade) [reply internal/campaigns/types.go]: 🤦‍♂

**Pull request merged**